realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp

```diff
@@ -609,33 +610,54 @@ JNIEXPORT jboolean JNICALL Java_io_realm_internal_SharedRealm_nativeCompact(JNIE
     TR_ENTER_PTR(shared_realm_ptr)
+    auto& shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
     try {
-        auto& shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
         auto* schema = reinterpret_cast<Schema*>(schema_ptr);
-        shared_realm->update_schema(*schema, static_cast<uint64_t>(version), nullptr, nullptr, true);
+        Realm::MigrationFunction migration_function = nullptr;
+        if (j_migration_callback) {
+            static JavaMethod run_migration_callback_method(env, j_shared_realm, "runMigrationCallback",
+                                                            "(Lio/realm/internal/SharedRealm$MigrationCallback;JJ)V");
+            migration_function = [&env, &j_shared_realm, &j_migration_callback, &shared_realm,
+                                  &version](SharedRealm old_realm, SharedRealm realm, Schema&) {
+                // We rely on the behaviour that Object Store will share_from_this from the original Realm and
+                // begin_transaction on it. So the realm passed from OS here will be the same realm instance which
+                // j_shared_realm is holding.
+                REALM_ASSERT_RELEASE(shared_realm == realm);
+                env->CallVoidMethod(j_shared_realm, run_migration_callback_method, j_migration_callback,
+                                    old_realm->schema_version(), version);
+            };
+        }
+        shared_realm->update_schema(*schema, static_cast<uint64_t>(version), migration_function, nullptr, true);
     }
-    CATCH_STD()
-}
-
-JNIEXPORT jboolean JNICALL Java_io_realm_internal_SharedRealm_nativeRequiresMigration(JNIEnv* env, jclass,
-                                                                                      jlong nativePtr,
-                                                                                      jlong nativeSchemaPtr)
-{
-
-    TR_ENTER()
-    try {
-        auto& shared_realm = *(reinterpret_cast<SharedRealm*>(nativePtr));
-        auto* schema = reinterpret_cast<Schema*>(nativeSchemaPtr);
-        const std::vector<SchemaChange>& change_list = shared_realm->schema().compare(*schema);
-        return static_cast<jboolean>(!change_list.empty());
+    catch (SchemaMismatchException& e) {
+        // An exception has been thrown in the migration block.
+        if (env->ExceptionCheck()) {
+            return;
+        }
+        static JavaClass migration_needed_class(env, JavaExceptionDef::RealmMigrationNeeded);
+        static JavaMethod constructor(env, migration_needed_class, "<init>",
+                                      "(Ljava/lang/String;Ljava/lang/String;)V");
+
+        jstring message = to_jstring(env, e.what());
+        jstring path = to_jstring(env, shared_realm->config().path);
+        jobject migration_needed_exception = env->NewObject(migration_needed_class, constructor, path, message);
+        env->Throw(reinterpret_cast<jthrowable>(migration_needed_exception));
+    }
```

Comment:
Actually here is the special case for converting. We need it to throw `MigrationNeededException` only here.